### DISCUSSION TITLE
Initial router.query is empty when URL query is set

### DISCUSSION THREAD

**Discussion by Regaddi** [vercel/next.js#13220] [Help]
# Bug report

## Describe the bug

When using the parsed `query` object from `useRouter` the initial value is always `{}`.
When a next.js app is opened with URL query parameters it should instead return the correct query, otherwise some IMO unnecessary state logic has to be built around.
I would imagine that it should be possible to read and parse the URL query before first render, so that `useRouter` can expose the correct query on first render.

## To Reproduce

Steps to reproduce the behavior, please provide code snippets or a repository:

- Go to https://4g5yg.sse.codesandbox.io/?abc=def
- Open Browser DevTools console
- You should see 3 logged items:
  - `initial query: {}`
  - `subsequent query: {}`
  - `subsequent query: {abc: "def"}`

The corresponding codesandbox is here: https://codesandbox.io/s/hungry-edison-4g5yg

## Expected behavior

The initial query should already contain `{abc:"def"}`

## System information

- OS: any
- Browser: any
- Version of Next.js: 9.4.2
- Version of Node.js: probably doesn't matter

## Additional context

I'll have to read a token from the URL when the application launches on any route and decide (based on the token and its validity) wether or not a redirect should happen. I could solve this with some additional state (something like `isTokenCheckPending`), but this is surely a workaround and doesn't compliment the overall ergonomics.

**Comment by illuminist** [comment]
> The query object will be empty during prerendering if the page is statically optimized.

https://nextjs.org/docs/api-reference/next/router#router-object

**Reply by Regaddi** [reply]
That's right, however this probably should not apply to the first client-side render (unless I misunderstand the meaning of "prerendering" and "first client-side render").

It's probably possible to work around this on my own by parsing window.location.search, but that's certainly not ergonomic when working with next/router.
Although the docs mention prerendering if the page is statically optimized I think it simply returns wrong information in this case.

**Reply by illuminist** [reply]
First client rendered page should match with prerendered page for statically optimized. Statically optimized prerendered page is created when building next app for production, and in that time, search query is undeterminable so it need to render the page with empty query. If you try to alter client side to render with query on first render, some element will not match and it will cause react some performance to hydrate the existing DOM with different elements.
